src/yadmark/benchmark/external/apfelpy_utils.py

```diff
@@ -0,0 +1,168 @@
+"""Utilities to help run APFEL++."""
+import numpy as np
+import apfelpy as ap
+
+
+def compute_apfelpy_data(theory, observables, pdf):
+    """Run APFEL++ to compute observables.
+
+    Parameters
+    ----------
+    theory : dict
+        theory runcard
+    observables : dict
+        observables runcard
+    pdf : lahapdf_like
+        PDF set
+
+    Returns
+    -------
+    dict
+        AFPEL numbers
+    """
+    ap.Banner()
+
+    # setup APFEL
+    # TODO: can we use other grids??
+    xgrid = ap.Grid(
+        [
+            ap.SubGrid(100, 1e-5, 3),
+            ap.SubGrid(60, 1e-1, 3),
+            ap.SubGrid(50, 6e-1, 3),
+            ap.SubGrid(50, 8e-1, 3),
+        ]
+    )
+    nQ = 50
+    QMin = 1
+    QMax = 200
+
+    # TODO: are we passing thr correctly??
+    thrs = [0, 0, 0, theory["mc"], theory["mb"], theory["mt"]]
+
+    fns = theory["FNS"]
+    if fns != "ZM-VFNS":
+        raise ValueError("APFEL++ benchmark contains only ZM-VFNS for the time being.")
+    pto = theory["PTO"]
+
+    alphas = ap.AlphaQCD(theory["alphas"], theory["Qref"], thrs, pto)
+    alphas = ap.TabulateObject(alphas, 2 * nQ, QMin, QMax, 3)
+
+    # mapping observables names to APFEL++ methods
+    init = ap.initializers
+    if observables["prDIS"] == "CC":
+        # TODO: CC are not running ... CKM matrix error
+        projectile_pids = {
+            "electron": 11,
+            "positron": -11,
+            "neutrino": 12,
+            "antineutrino": -12,
+        }
+        if projectile_pids[observables["ProjectileDIS"]] > 0:
+            apfelpy_structure_functions = {
+                "F2": init.InitializeF2CCMinusObjectsZM,
+                "FL": init.InitializeFLCCMinusObjectsZM,
+                "F3": init.InitializeF3CCMinusObjectsZM,
+            }
+        else:
+            apfelpy_structure_functions = {
+                "F2": init.InitializeF2CCPlusObjectsZM,
+                "FL": init.InitializeFLCCPlusObjectsZM,
+                "F3": init.InitializeF3CCPlusObjectsZM,
+            }
+    else:
+        apfelpy_structure_functions = {
+            "F2": init.InitializeF2NCObjectsZM,
+            "FL": init.InitializeFLNCObjectsZM,
+            "F3": init.InitializeF3NCObjectsZM,
+            "g1": init.Initializeg1NCObjectsZM,
+            "gL": init.InitializegLNCObjectsZM,
+            "g4": init.Initializeg4NCObjectsZM,
+        }
+
+    # TODO: do you need to disable evolution also here?
+    # disable APFEL++ evolution: we are interested in the pure DIS part
+    dglapobj = ap.initializers.InitializeDglapObjectsQCD(xgrid, thrs)
+    # Construct the DGLAP objects
+    evolvedPDFs = ap.builders.BuildDglap(
+        dglapobj, ap.utilities.LHToyPDFs, theory["Q0"], pto, alphas.Evaluate
```

Comment:
Here I think we should pass `pdf.xfxQ`, but somehow Apfel++ was complaining about the banana object..